### PR TITLE
Allow user to set a propagation policy on deletion

### PR DESCRIPTION
**Pull request by xguerin**

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

**Comment by rohanKanojia** [comment]
@xguerin : looks like there are some compilations issues in openshift-client/ module. Could you please update that module also according to your changes?

**Comment by xguerin** [comment]
@rohanKanojia Apologies for the partial patch, it should be fixed now.

**Comment by xguerin** [comment]
@rohanKanojia Not quite sure if the remaining failure is linked to my patch or not.

**Comment by rohanKanojia** [comment]
@xguerin : yeah, right. Let me restart it.

**Comment by xguerin** [comment]
Sure thing. I'll update the CHANGELOG as well.

**Comment by xguerin** [comment]
@rohanKanojia Done.

**Comment by rohanKanojia** [comment]
retest this please

**Comment by xguerin** [comment]
@rohanKanojia All good on my end 👍

**Comment by rohanKanojia** [comment]
[merge]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/DefaultKubernetesClient.java [hunk 1]
```diff
@@ -134,20 +134,20 @@ public static DefaultKubernetesClient fromConfig(InputStream is) {
...
-    return new NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl(httpClient, getConfiguration(), getNamespace(), null, false, false, new ArrayList<Visitor>(), s, null, null, -1, true) {
+    return new NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl(httpClient, getConfiguration(), getNamespace(), null, false, false, new ArrayList<Visitor>(), s, null, null, -1, null, false) {
     };
   }
 
 
   @Override
   public NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicable<HasMetadata, Boolean> resource(HasMetadata item) {
-    return new NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl(httpClient, getConfiguration(), getNamespace(), null, false, false, new ArrayList<Visitor>(), item, -1, true) {
+    return new NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl(httpClient, getConfiguration(), getNamespace(), null, false, false, new ArrayList<Visitor>(), item, -1, null, false) {
```
[the remark this answers is not in the document]
Comment (xguerin) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/DefaultKubernetesClient.java]: My apologies, This is a mistake indeed. I must have made that change once i realized that either the propagation policy or the cascading flag may be set, but not both at the same time. I’ll submit a patch to restore that value to true.

[the remark this answers is not in the document]
Comment (rohanKanojia) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/DefaultKubernetesClient.java]: no, worries. I have prepared a patch to revert this to false.

**Comment by soumava** [comment]
I'm trying to emulate kubectl's payload from the API for a forced pod deletion (--grace-period=0 --force). With -v=8 I see the following payload:
```
I1221 09:09:49.860340   49454 request.go:968] Request Body: {"gracePeriodSeconds":0,"propagationPolicy":"Background"}
I1221 09:09:49.860399   49454 round_trippers.go:420] DELETE https://<kube-api-server>:443/api/v1/namespaces/<namespace>/pods/<pod-name>
```

When I try to use the fabric8 client to set both propagationPolicy and gracePeriod it doesn't seem possible (without casting one of the intermediate results), as the methods `.withPropagationPolicy()` and `.withGracePeriod()` both return a Deletable<Boolean>. 

```
client.pods()
                    .inNamespace(<namespace>)
                    .withName(<pod-name>)
                    .withPropagationPolicy("Background")
                    .withGracePeriod(0)             <---------- doesn't work
```
```
val operation = client.pods()
                    .inNamespace(<namespace>)
                    .withName(<pod-name>)
                    .withPropagationPolicy("Background")

val deleted = (operation as GracePeriodConfigurable<Deletable<Boolean>>)
                    .withGracePeriod(0)
                    .delete()
```

Is there a way to address this that would not require the caller to cast?

**Comment by rohanKanojia** [comment]
I think you're using Kotlin. Seems like `withGracePeriod` and `withPropagationPolicy` can't be used together with current implementation. You can use `cascading=true` flag which sets `orphanDependents=false` (if that solves your us case)
```
      Boolean deleted = client.pods().inNamespace("default").withName("foo")
        .cascading(true)
        .withGracePeriod(0)
        .delete();
```

**Pull request merged**